src/mame/machine/k573fpga.cpp

```diff
@@ -0,0 +1,592 @@
+        mp3_last_decrypt_adr = mp3_start_adr;
+        last_copied_samples = 0;
+        last_buffer_size_channel_l = 0;
+        last_buffer_size_channel_r = 0;
+
+        last_position_update = 0;
+        position_diff = 0;
+        decrypt_finished = false;
+
+        if (channel_l_pcm) {
+            free(channel_l_pcm);
+            channel_l_pcm = nullptr;
+        }
+
+        if (channel_r_pcm) {
+            free(channel_r_pcm);
+            channel_r_pcm = nullptr;
+        }
+    }
+
+    if ((data & 0xe000) == 0xe000) {
+        mp3_decrypt_mode = true;
+    } else {
+        mp3_decrypt_mode = false;
+    }
+}
+
+int32_t k573fpga_device::find_enc_key()
+{
+    if (k573enc_lookup.find(crypto_key1) == k573enc_lookup.end()) {
+        return -1;
+    }
+
+    return k573enc_lookup[crypto_key1];
+}
+
+
+inline uint8_t IsBitSet(int value, int n)
+{
+    return (value >> n) & 1;
+}
+
+inline uint16_t BitSwap16(int v, int b15, int b14, int b13, int b12, int b11, int b10, int b9, int b8,
+    int b7, int b6, int b5, int b4, int b3, int b2, int b1, int b0)
+{
+    return (IsBitSet(v, b15) << 15) |
+        (IsBitSet(v, b14) << 14) |
+        (IsBitSet(v, b13) << 13) |
+        (IsBitSet(v, b12) << 12) |
+        (IsBitSet(v, b11) << 11) |
+        (IsBitSet(v, b10) << 10) |
+        (IsBitSet(v, b9) << 9) |
+        (IsBitSet(v, b8) << 8) |
+        (IsBitSet(v, b7) << 7) |
+        (IsBitSet(v, b6) << 6) |
+        (IsBitSet(v, b5) << 5) |
+        (IsBitSet(v, b4) << 4) |
+        (IsBitSet(v, b3) << 3) |
+        (IsBitSet(v, b2) << 2) |
+        (IsBitSet(v, b1) << 1) |
+        (IsBitSet(v, b0) << 0);
+}
+
+inline uint16_t k573fpga_device::fpga_decrypt_byte_real(uint16_t v)
+{
+    uint16_t m = crypto_key1 ^ crypto_key2;
+
+    v = BitSwap16(
+        v,
+        15 - IsBitSet(m, 0xF),
+        14 + IsBitSet(m, 0xF),
+        13 - IsBitSet(m, 0xE),
+        12 + IsBitSet(m, 0xE),
+        11 - IsBitSet(m, 0xB),
+        10 + IsBitSet(m, 0xB),
+        9 - IsBitSet(m, 0x9),
+        8 + IsBitSet(m, 0x9),
+        7 - IsBitSet(m, 0x8),
+        6 + IsBitSet(m, 0x8),
+        5 - IsBitSet(m, 0x5),
+        4 + IsBitSet(m, 0x5),
+        3 - IsBitSet(m, 0x3),
+        2 + IsBitSet(m, 0x3),
+        1 - IsBitSet(m, 0x2),
+        0 + IsBitSet(m, 0x2)
+    );
+
+    v ^= (IsBitSet(m, 0xD) << 14) ^
+        (IsBitSet(m, 0xC) << 12) ^
+        (IsBitSet(m, 0xA) << 10) ^
+        (IsBitSet(m, 0x7) << 8) ^
+        (IsBitSet(m, 0x6) << 6) ^
+        (IsBitSet(m, 0x4) << 4) ^
+        (IsBitSet(m, 0x1) << 2) ^
+        (IsBitSet(m, 0x0) << 0);
+
+    v ^= BitSwap16(
+        crypto_key3,
+        7, 0, 6, 1,
+        5, 2, 4, 3,
+        3, 4, 2, 5,
+        1, 6, 0, 7
+    );
+
+    return (v >> 8) | ((v & 0xff) << 8);
+}
+
+inline uint16_t k573fpga_device::fpga_decrypt_byte_fake(uint16_t data, uint32_t crypto_idx)
+{
+    // Not the real algorithm. Only used for DDR Solo Bass Mix (ddrsbm)
+    int32_t crypto_enc_idx = find_enc_key();
```

Comment:
Do you mean that ddrsbm has a different fpga firmware?

Reply:
I've elaborated in [this comment](https://github.com/mamedev/mame/pull/5055#discussion_r284123984), but yes, ddrsbm uses a completely different FPGA firmware from everything else. Every other game seems to use an FPGA firmware derived at the very least from a standard program, and then ddrsbm comes in and does something completely different.